_posts/2020-12-27-spring-boot-exception-hanlding.md

```diff
@@ -0,0 +1,596 @@
+Giving stack trace in the error response can save our Developers and QA's trouble of crawling through the log files or console logs or by
+whatever means we have used to expose our logs. As we saw in the [Spring Boot's Default Exception Handling Mechanism](#spring-boots-default-exception-handling-mechanism) section that Spring was already providing us
+with this functionality but now, as we are handling error responses this also needs to be handled by ourselves.
+
+To achieve this, similar to `server.error` first we have introduced a property named `reflectoring.trace` which if set to `true`
+will enable `stack_trace` in the response. In order to actually get `stackTrace` in an API response, our clients must pass
+`trace` parameter with the value `true`.
+
+```text
+curl --location --request GET 'http://localhost:8080/product/1?trace=true'
+```
+
+Now, as the behavior of `strace_trace` is controlled by our feature flag in our properties file, we can remove it or set
+to `false` when we deploy in production environments.
+
+### Catch-All Exception
+
+*Gotta catch em all*
+
+```java
+try{
+    performSomeOperation();
+} catch(OperationSpecificException ex){
+    //...
+} catch(Exception catchAllExcetion){
+    //...    
+}
+```
+
+As a cautionary measure, we often surround our top level method's body with a catch-all Try-Catch exception handler block
+in order to avoid any unwanted side effects or behavior. The `handleAllUncaughtException` method in our controller behaves
+similarly. **It will catch all the exceptions for which a handler method doesn't exist**.
+
+One thing I would like to note here is that even if we don't have this catch-all exception handler, Spring will handle it
+anyway. We handled it because we wanted the response to go in our format rather than Spring's. 
+
+Handler method might also be a good place to log exceptions. This catch-all is one those exceptions that we must log, as 
+they might give insight into a possible bug. While we can skip logging field validation exceptions such as `MethodArgumentNotValidException`
+as they are raised because of syntactically invalid input.
+
+
+### Order Of Exception Handler Method
+
+The order in which you mention the handler methods doesn't matter. **Spring will first look for the specific exception handler method**.
+If it fails to find it then it will look for its parent class which in our case is `RuntimeException` and hence
+`handleAllUncaughtException` method will finally handle the exception.
+
+This should help us handle the exceptions in this particular controller, but what if these same exceptions are being thrown
+by other controllers too? How do we handle those? Do we create the same handlers in all controllers or create a base class with
+common handlers and extend it in all controllers?
+
+Luckily, we don't have to do any of that. Spring provides a very elegant solution to this problem in form of Controller Advices
+let's study them.
+
+## `@ControllerAdvice`
```

Comment:
Perhaps add an info box in this section that explains why it's called an "advice" (from AOP)?